genie-agent/src/main/java/com/netflix/genie/agent/jobexecution/statemachine/StateMachineConfig.java

```diff
@@ -0,0 +1,207 @@
+        for (JobExecutionListener listener : listeners) {
+            stateMachine.addStateListener(listener);
+            for (State<States, Events> state : stateMachine.getStates()) {
+                state.addActionListener(listener);
+            }
+        }
+
+        return stateMachine;
+    }
+
+    @Bean
+    @Lazy
+    Collection<Pair<States, StateAction>> statesWithActions(
+        final StateAction.Initialize initializeAction,
+        final StateAction.ConfigureAgent configureAgentAction,
+        final StateAction.ConfigureJob configureJobAction,
+        final StateAction.SetUpJob setUpJobAction,
+        final StateAction.LaunchJob launchJobAction,
+        final StateAction.MonitorJob monitorJobAction,
+        final StateAction.CleanupJob cleanupJobAction,
+        final StateAction.Shutdown shutdownAction,
+        final StateAction.HandleError handleErrorAction
+    ) {
+        return Arrays.asList(
+            Pair.of(States.INITIALIZE, initializeAction),
+            Pair.of(States.CONFIGURE_AGENT, configureAgentAction),
+            Pair.of(States.CONFIGURE_JOB, configureJobAction),
+            Pair.of(States.SETUP_JOB, setUpJobAction),
+            Pair.of(States.LAUNCH_JOB, launchJobAction),
+            Pair.of(States.MONITOR_JOB, monitorJobAction),
+            Pair.of(States.CLEANUP_JOB, cleanupJobAction),
+            Pair.of(States.SHUTDOWN, shutdownAction),
+            Pair.of(States.HANDLE_ERROR, handleErrorAction)
+        );
+    }
+
+    @Bean
+    @Lazy
+    Collection<Triple<States, Events, States>> eventDrivenTransitions() {
+        return Arrays.asList(
+            // Regular execution
+            Triple.of(States.READY, Events.START, States.INITIALIZE),
+            Triple.of(States.INITIALIZE, Events.INITIALIZE_COMPLETE, States.CONFIGURE_AGENT),
+            Triple.of(States.CONFIGURE_AGENT, Events.CONFIGURE_AGENT_COMPLETE, States.CONFIGURE_JOB),
+            Triple.of(States.CONFIGURE_JOB, Events.CONFIGURE_JOB_COMPLETE, States.SETUP_JOB),
+            Triple.of(States.SETUP_JOB, Events.SETUP_JOB_COMPLETE, States.LAUNCH_JOB),
+            Triple.of(States.LAUNCH_JOB, Events.LAUNCH_JOB_COMPLETE, States.MONITOR_JOB),
+            Triple.of(States.MONITOR_JOB, Events.MONITOR_JOB_COMPLETE, States.CLEANUP_JOB),
+            Triple.of(States.CLEANUP_JOB, Events.CLEANUP_JOB_COMPLETE, States.SHUTDOWN),
+            Triple.of(States.SHUTDOWN, Events.SHUTDOWN_COMPLETE, States.END),
+            // Job cancellation
+            Triple.of(States.READY, Events.CANCEL_JOB_LAUNCH, States.CLEANUP_JOB),
+            Triple.of(States.INITIALIZE, Events.CANCEL_JOB_LAUNCH, States.CLEANUP_JOB),
+            Triple.of(States.CONFIGURE_AGENT, Events.CANCEL_JOB_LAUNCH, States.CLEANUP_JOB),
+            Triple.of(States.CONFIGURE_JOB, Events.CANCEL_JOB_LAUNCH, States.CLEANUP_JOB),
+            Triple.of(States.SETUP_JOB, Events.CANCEL_JOB_LAUNCH, States.CLEANUP_JOB)
+        );
+    }
+
+    @Bean
+    @Lazy
```

Comment:
All beans are lazy because we don't want to initialize the state machine (and the million depending beans for other commands, such as ping, info, etc).